### PR TITLE
Fix TSAN Race in Overriding poll_function

### LINKED ISSUE #14388: TSAN failure in AsyncEnd2end/AsyncEnd2endTest.ShutdownThenWait/15 GRPC_POLL_STRATEGY=poll
Happens about 1% of the time

https://sponge.corp.google.com/target?id=b5cedbcc-ebd1-4f2b-abfd-b9911e6aee91&target=github/grpc/cpp_linux_tsan_native&searchFor=&show=ALL&sortBy=STATUS

https://sponge.corp.google.com/target?id=51ef4728-3f82-47c0-bcaf-1c065c0e06cd&target=github/grpc/cpp_linux_tsan_native&searchFor=&show=ALL&sortBy=STATUS

Test code that overrides poll_funtion is racing with pollset_work, which is actually calling that function

### PR DESCRIPTION
**Pull request by ncteisen**
Fixes #14388

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: test/cpp/end2end/async_end2end_test.cc [hunk 1]
```diff
@@ -77,14 +77,17 @@ static int maybe_assert_non_blocking_poll(struct pollfd* pfds, nfds_t nfds,
 class PollOverride {
  public:
   PollOverride(grpc_poll_function_type f) {
-    prev_ = grpc_poll_function;
-    grpc_poll_function = f;
+    gpr_atm pf = (gpr_atm)&grpc_poll_function;
```
Comment (vjpai) [inline test/cpp/end2end/async_end2end_test.cc]: Please use C++ style casts

File: test/cpp/end2end/async_end2end_test.cc [hunk 2]
```diff
@@ -77,14 +77,17 @@ static int maybe_assert_non_blocking_poll(struct pollfd* pfds, nfds_t nfds,
...
+    prev_ = gpr_atm_full_xchg(&pf, (gpr_atm)f);
   }
 
-  ~PollOverride() { grpc_poll_function = prev_; }
+  ~PollOverride() {
+    gpr_atm pf = (gpr_atm)&grpc_poll_function;
```
Comment (vjpai) [inline test/cpp/end2end/async_end2end_test.cc]: Likewise

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by vjpai** [comment]
Broader question: is this the right approach? AFAICT, polling is overridden at setup and goes away at test destruction. Doesn't the fact that there is a race indicate that some thread is still active at the time of test destruction (which would be a bad thing)?

**Comment by vjpai** [comment]
Just saw the stack trace and saw that the other thread that is active is a timer thread. It seems to me that this whole concept that we have needs a revisit. The thing is that this PollingOverride test was created before we had a dedicated timer thread, and I think it's making the previously-valid assumption that there aren't lurking library threads doing stuff behind the test's back. I'm not that comfortable with atomically switching a poll function while some other thread is in the middle of polling with a different polling function.

File: test/cpp/end2end/async_end2end_test.cc [hunk 3]
```diff
@@ -77,14 +77,17 @@ static int maybe_assert_non_blocking_poll(struct pollfd* pfds, nfds_t nfds,
 class PollOverride {
  public:
   PollOverride(grpc_poll_function_type f) {
-    prev_ = grpc_poll_function;
-    grpc_poll_function = f;
+    gpr_atm pf = reinterpret_cast<gpr_atm>(&grpc_poll_function);
+    prev_ = gpr_atm_full_xchg(&pf, reinterpret_cast<gpr_atm>(f));
```
Comment (vjpai) [inline test/cpp/end2end/async_end2end_test.cc]: Ok, expressing this as reinterpret_cast's makes it much clearer, IMO. What is this supposed to be doing? You're exchanging into a local variable, right? Shouldn't this be something like

``` cpp
prev_ = gpr_atm_full_xchg(reinterpret_cast<gpr_atm*>(&grpc_poll_function), reinterpret_cast<gpr_atm>(f));
```

Comment (vjpai) [reply test/cpp/end2end/async_end2end_test.cc]: My previous comment was on the code, but I still have the broader question about what the right approach should be.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by ncteisen** [comment]
Done

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by ncteisen** [comment]
Got a green run earlier, merging

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**